File: gma-next-gen-sdk-android-examples-main/java/nextgenexample/app/src/main/java/com/example/nextgenexample/nativead/NativeFragment.java
/*
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.nextgenexample.nativead;

import android.app.Activity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.example.nextgenexample.AdFragment;
import com.example.nextgenexample.Constant;
import com.example.nextgenexample.R;
import com.example.nextgenexample.databinding.FragmentNativeBinding;
import com.example.nextgenexample.databinding.NativeAdBinding;
import com.google.android.libraries.ads.mobile.sdk.common.FullScreenContentError;
import com.google.android.libraries.ads.mobile.sdk.common.LoadAdError;
import com.google.android.libraries.ads.mobile.sdk.common.VideoController;
import com.google.android.libraries.ads.mobile.sdk.common.VideoOptions;
import com.google.android.libraries.ads.mobile.sdk.nativead.MediaContent;
import com.google.android.libraries.ads.mobile.sdk.nativead.NativeAd;
import com.google.android.libraries.ads.mobile.sdk.nativead.NativeAd.NativeAdType;
import com.google.android.libraries.ads.mobile.sdk.nativead.NativeAdEventCallback;
import com.google.android.libraries.ads.mobile.sdk.nativead.NativeAdLoader;
import com.google.android.libraries.ads.mobile.sdk.nativead.NativeAdLoaderCallback;
import com.google.android.libraries.ads.mobile.sdk.nativead.NativeAdRequest;
import com.google.android.libraries.ads.mobile.sdk.nativead.NativeAdView;
import java.util.List;

/** An [AdFragment] subclass that loads a native ad. */
public class NativeFragment extends AdFragment<FragmentNativeBinding> {
  // Default constructor required for fragment instantiation.
  public NativeFragment() {}

  // Sample native image ad unit ID.
  private static final String IMAGE_AD_UNIT_ID = "ca-app-pub-3940256099942544/2247696110";
  // Sample native video ad unit ID.
  private static final String VIDEO_AD_UNIT_ID = "ca-app-pub-3940256099942544/1044960115";

  private NativeAd lastNativeAd;
  private boolean isUIEnabled = true;
  private CustomVideoControlsView customControls;

  @Override
  protected BindingInflater<FragmentNativeBinding> getBindingInflater() {
    return FragmentNativeBinding::inflate;
  }

  @Override
  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);

    binding.refreshAdButton.setOnClickListener(
        unusedView -> {
          loadAd();
        });
    binding.checkRequestVideo.setOnClickListener(
        unusedView -> {
          updateUI();
        });
  }

  @Override
  public void onDestroyView() {
    super.onDestroyView();

    // Always call destroy() on ads on removal.
    destroyNativeAd();
    }

  private void destroyNativeAd() {
    if (lastNativeAd != null) {
      lastNativeAd.destroy();
      lastNativeAd = null;
    }
    customControls = null;
  }

  private void loadAd() {
    setUIEnabled(false);

    // Build an ad request with native ad options to customize the user experience.
    String adUnitID = binding.checkRequestVideo.isChecked() ? VIDEO_AD_UNIT_ID : IMAGE_AD_UNIT_ID;

    // Build an ad request with native ad options to customize the user experience.
    VideoOptions videoOptions =
        new VideoOptions.Builder().setStartMuted(binding.checkStartMuted.isChecked()).build();

    NativeAdRequest adRequest =
        new NativeAdRequest.Builder(adUnitID, List.of(NativeAdType.NATIVE))
            .setVideoOptions(videoOptions)
            .build();

    // Define the callback to handle successful or failed ad loading.
    NativeAdLoaderCallback adLoaderCallback =
        new NativeAdLoaderCallback() {
          @Override
          public void onNativeAdLoaded(@NonNull NativeAd nativeAd) {
            Log.d(Constant.TAG, "Native ad loaded.");
            showToast("Native ad loaded.");
            runOnUiThread(
                () -> {
                  // Remove all old ad views when loading a new native ad.
                  binding.nativeViewContainer.removeAllViews();
                  // Always call destroy() on ads on removal.
                  destroyNativeAd();
                  setNativeAdEventCallback(nativeAd);
                  displayNativeAd(nativeAd);
                  lastNativeAd = nativeAd;
                  setUIEnabled(true);
                });
          }

          @Override
          public void onAdFailedToLoad(@NonNull LoadAdError adError) {
            Log.d(Constant.TAG, "Native ad failed to load :" + adError);
            runOnUiThread(
                () -> {
                  showToast("Native ad failed to load with error code: " + adError.getCode());
                  setUIEnabled(true);
                });
          }
        };

    // Load the native ad with the ad request and callback.
    NativeAdLoader.load(adRequest, adLoaderCallback);
  }

  private void setNativeAdEventCallback(NativeAd nativeAd) {
    nativeAd.setAdEventCallback(
        new NativeAdEventCallback() {
          @Override
          public void onAdShowedFullScreenContent() {
            Log.d(Constant.TAG, "Native ad showed full screen content.");
          }

          @Override
          public void onAdDismissedFullScreenContent() {
            Log.d(Constant.TAG, "Native ad dismissed full screen content.");
          }

          @Override
          public void onAdFailedToShowFullScreenContent(
              @NonNull FullScreenContentError fullScreenContentError) {
            Log.d(
                Constant.TAG,
                "Native ad failed to show full screen content with error : "
                    + fullScreenContentError);
          }

          @Override
          public void onAdImpression() {
            Log.d(Constant.TAG, "Native ad recorded an impression.");
          }

          @Override
          public void onAdClicked() {
            Log.d(Constant.TAG, "Native ad recorded a click.");
          }
        });
  }

  private void displayNativeAd(NativeAd nativeAd) {
    // Inflate the native ad view and add to the active view hierarchy.
    NativeAdBinding nativeAdBinding = NativeAdBinding.inflate(getLayoutInflater());
    NativeAdView nativeAdView = nativeAdBinding.getRoot();
    binding.nativeViewContainer.addView(nativeAdView);

    // Populate the view elements with their respective native ad asset.
    nativeAdBinding.adAdvertiser.setText(nativeAd.getAdvertiser());
    nativeAdBinding.adBody.setText(nativeAd.getBody());
    nativeAdBinding.adCallToAction.setText(nativeAd.getCallToAction());
    nativeAdBinding.adHeadline.setText(nativeAd.getHeadline());
    nativeAdBinding.adPrice.setText(nativeAd.getPrice());
    nativeAdBinding.adStore.setText(nativeAd.getStore());
    if (nativeAd.getIcon() != null && nativeAd.getIcon().getDrawable() != null) {
      nativeAdBinding.adIcon.setImageDrawable(nativeAd.getIcon().getDrawable());
    }
    if (nativeAd.getStarRating() != null) {
      float starRating = nativeAd.getStarRating().floatValue();
      nativeAdBinding.adStars.setRating(starRating);
    }

    // Hide views for assets that don't have data.
    nativeAdBinding.adAdvertiser.setVisibility(
        (nativeAd.getAdvertiser() == null ? View.INVISIBLE : View.VISIBLE));
    nativeAdBinding.adPrice.setVisibility(
        (nativeAd.getPrice() == null ? View.INVISIBLE : View.VISIBLE));
    nativeAdBinding.adStore.setVisibility(
        (nativeAd.getStore() == null ? View.INVISIBLE : View.VISIBLE));
    nativeAdBinding.adIcon.setVisibility(
        (nativeAd.getIcon() == null ? View.INVISIBLE : View.VISIBLE));
    nativeAdBinding.adStars.setVisibility(
        (nativeAd.getStarRating() == null ? View.INVISIBLE : View.VISIBLE));

    // Map each asset view property to the corresponding view in your view hierarchy.
    nativeAdView.setAdvertiserView(nativeAdBinding.adAdvertiser);
    nativeAdView.setBodyView(nativeAdBinding.adBody);
    nativeAdView.setCallToActionView(nativeAdBinding.adCallToAction);
    nativeAdView.setHeadlineView(nativeAdBinding.adHeadline);
    nativeAdView.setPriceView(nativeAdBinding.adPrice);
    nativeAdView.setStoreView(nativeAdBinding.adStore);
    nativeAdView.setIconView(nativeAdBinding.adIcon);
    nativeAdView.setStarRatingView(nativeAdBinding.adStars);

    // Inform the Google Mobile Ads SDK that you have finished populating the native ad
    // views with this native ad.
    nativeAdView.registerNativeAd(nativeAd, nativeAdBinding.adMedia);

    MediaContent mediaContent = nativeAd.getMediaContent();
    VideoController videoController = mediaContent.getVideoController();
    if (videoController != null && mediaContent.getHasVideoContent()) {
      // If the main asset is a video, set the videoLifecycleCallbacks.
      binding.textVideoStatus.setText(getString(R.string.nativead_video_play));
      videoController.setVideoLifecycleCallbacks(getVideoCallbacks());
    } else {
      // The main asset is an image
      binding.textVideoStatus.setText(getString(R.string.nativead_video_none));
    }
  }

  private VideoController.VideoLifecycleCallbacks getVideoCallbacks() {
    return new VideoController.VideoLifecycleCallbacks() {
      @Override
      public void onVideoStart() {
        binding.textVideoStatus.setText(getString(R.string.nativead_video_started));
        if (customControls != null) {
          customControls.onVideoStart();
        }
      }

      @Override
      public void onVideoEnd() {
        // Publishers should allow native ads to complete video playback before
        // refreshing or replacing them with another ad in the same UI location.
        binding.textVideoStatus.setText(getString(R.string.nativead_video_ended));
        if (customControls != null) {
          customControls.onVideoEnd();
        }
      }

      @Override
      public void onVideoPause() {
        binding.textVideoStatus.setText(getString(R.string.nativead_video_pause));
        if (customControls != null) {
          customControls.onVideoPause();
        }
      }

      @Override
      public void onVideoPlay() {
        binding.textVideoStatus.setText(getString(R.string.nativead_video_play));
        if (customControls != null) {
          customControls.onVideoPlay();
        }
      }

      @Override
      public void onVideoMute(boolean muted) {
        if (customControls != null) {
          customControls.onVideoMute(muted);
        }
      }
    };
  }

  private void setUIEnabled(boolean enableUI) {
    isUIEnabled = enableUI;
    updateUI();
  }

  private void updateUI() {
    runOnUiThread(
        () -> {
          binding.refreshAdButton.setEnabled(isUIEnabled);
          binding.checkRequestVideo.setEnabled(isUIEnabled);
          boolean videoUIEnabled = isUIEnabled && binding.checkRequestVideo.isChecked();
          binding.checkStartMuted.setEnabled(videoUIEnabled);
        });
  }

  public void runOnUiThread(Runnable action) {
    Activity activity = getActivity();
    if (activity == null) {
      return;
    }
    activity.runOnUiThread(action);
  }
}

File: gma-next-gen-sdk-android-examples-main/kotlin/nextgenexample/app/src/main/java/com/example/next_gen_example/native/NativeFragment.kt
/*
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.next_gen_example.native

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.example.next_gen_example.AdFragment
import com.example.next_gen_example.Constant
import com.example.next_gen_example.R
import com.example.next_gen_example.databinding.FragmentNativeBinding
import com.example.next_gen_example.databinding.NativeAdBinding
import com.google.android.libraries.ads.mobile.sdk.common.FullScreenContentError
import com.google.android.libraries.ads.mobile.sdk.common.LoadAdError
import com.google.android.libraries.ads.mobile.sdk.common.VideoController
import com.google.android.libraries.ads.mobile.sdk.common.VideoOptions
import com.google.android.libraries.ads.mobile.sdk.nativead.NativeAd
import com.google.android.libraries.ads.mobile.sdk.nativead.NativeAdEventCallback
import com.google.android.libraries.ads.mobile.sdk.nativead.NativeAdLoader
import com.google.android.libraries.ads.mobile.sdk.nativead.NativeAdLoaderCallback
import com.google.android.libraries.ads.mobile.sdk.nativead.NativeAdRequest
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/** A simple [Fragment] subclass that loads a native ad. */
class NativeFragment : AdFragment<FragmentNativeBinding>() {
  override val bindingInflater: (LayoutInflater, ViewGroup?, Boolean) -> FragmentNativeBinding
    get() = FragmentNativeBinding::inflate

  private var lastNativeAd: NativeAd? = null
  private var customControls: CustomVideoControlsView? = null
  private var isUIEnabled = true

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    binding.refreshAdButton.setOnClickListener { loadAd() }
    binding.checkRequestVideo.setOnClickListener { updateUI() }
  }

  override fun onDestroyView() {
    super.onDestroyView()
    // Always call destroy() on ads on removal.
    destroyNativeAd()
  }

  private fun destroyNativeAd() {
    lastNativeAd?.destroy()
    lastNativeAd = null
    customControls = null
  }

  private fun loadAd() {
    setUIEnabled(false)

    // Build an ad request with native ad options to customize the user experience.
    val adUnitID = if (binding.checkRequestVideo.isChecked) VIDEO_AD_UNIT_ID else IMAGE_AD_UNIT_ID
    val videoOptions: VideoOptions =
      VideoOptions.Builder().setStartMuted(binding.checkStartMuted.isChecked).build()
    val adRequest =
      NativeAdRequest.Builder(adUnitID, listOf(NativeAd.NativeAdType.NATIVE))
        .setVideoOptions(videoOptions)
        .build()

    // Define the callback to handle successful ad loading or failed ad loading.
    val adCallback =
      object : NativeAdLoaderCallback {
        override fun onNativeAdLoaded(nativeAd: NativeAd) {
          Log.d(Constant.TAG, "Native ad loaded.")
          CoroutineScope(Dispatchers.Main).launch {
            // Remove all old ad views when loading a new native ad.
            binding.nativeViewContainer.removeAllViews()
            // Always call destroy() on ads on removal.
            destroyNativeAd()
            setEventCallback(nativeAd)
            displayNativeAd(nativeAd)
            lastNativeAd = nativeAd
            setUIEnabled(true)
          }
        }

        override fun onAdFailedToLoad(adError: LoadAdError) {
          Log.e(Constant.TAG, "Native ad failed to load: $adError")
          showToast("Ad failed to load.")
          setUIEnabled(true)
        }
      }
    // Load the native ad with our request and callback.
    NativeAdLoader.load(adRequest, adCallback)
  }

  private fun setEventCallback(nativeAd: NativeAd) {
    nativeAd.adEventCallback =
      object : NativeAdEventCallback {
        override fun onAdShowedFullScreenContent() {
          Log.d(Constant.TAG, "Native ad showed full screen content.")
        }

        override fun onAdDismissedFullScreenContent() {
          Log.d(Constant.TAG, "Native ad dismissed full screen content.")
        }

        override fun onAdFailedToShowFullScreenContent(
          fullScreenContentError: FullScreenContentError
        ) {
          Log.d(
            Constant.TAG,
            "Native ad failed to show full screen content with error: $fullScreenContentError",
          )
        }

        override fun onAdImpression() {
          Log.d(Constant.TAG, "Native ad recorded an impression.")
        }

        override fun onAdClicked() {
          Log.d(Constant.TAG, "Native ad recorded a click.")
        }
      }
  }

  private fun displayNativeAd(nativeAd: NativeAd) {
    // Inflate the native ad view and add it to the view hierarchy.
    val nativeAdBinding = NativeAdBinding.inflate(layoutInflater)
    binding.nativeViewContainer.addView(nativeAdBinding.root)

    // Set the native ad view elements.
    val nativeAdView = nativeAdBinding.root
    nativeAdView.advertiserView = nativeAdBinding.adAdvertiser
    nativeAdView.bodyView = nativeAdBinding.adBody
    nativeAdView.callToActionView = nativeAdBinding.adCallToAction
    nativeAdView.headlineView = nativeAdBinding.adHeadline
    nativeAdView.iconView = nativeAdBinding.adIcon
    nativeAdView.priceView = nativeAdBinding.adPrice
    nativeAdView.starRatingView = nativeAdBinding.adStars
    nativeAdView.storeView = nativeAdBinding.adStore

    // Set the view element with the native ad assets.
    nativeAdBinding.adAdvertiser.text = nativeAd.advertiser
    nativeAdBinding.adBody.text = nativeAd.body
    nativeAdBinding.adCallToAction.text = nativeAd.callToAction
    nativeAdBinding.adHeadline.text = nativeAd.headline
    nativeAdBinding.adIcon.setImageDrawable(nativeAd.icon?.drawable)
    nativeAdBinding.adPrice.text = nativeAd.price
    nativeAd.starRating?.toFloat().also { value ->
      if (value != null) {
        nativeAdBinding.adStars.rating = value
      }
    }
    nativeAdBinding.adStore.text = nativeAd.store

    // Hide views for assets that don't have data.
    nativeAdBinding.adAdvertiser.visibility =
      if (nativeAd.advertiser == null) View.INVISIBLE else View.VISIBLE
    nativeAdBinding.adIcon.visibility = if (nativeAd.icon == null) View.INVISIBLE else View.VISIBLE
    nativeAdBinding.adPrice.visibility =
      if (nativeAd.price == null) View.INVISIBLE else View.VISIBLE
    nativeAdBinding.adStars.visibility =
      if (nativeAd.starRating == null) View.INVISIBLE else View.VISIBLE
    nativeAdBinding.adStore.visibility =
      if (nativeAd.store == null) View.INVISIBLE else View.VISIBLE

    // Inform the Google Mobile Ads SDK that you have finished populating the native ad views
    // with this native ad.
    nativeAdView.registerNativeAd(nativeAd, nativeAdBinding.adMedia)

    val videoController = nativeAd.mediaContent.videoController
    if (videoController != null && nativeAd.mediaContent.hasVideoContent) {
      // If the main asset is a video, set the videoLifecycleCallbacks.
      binding.textVideoStatus.text = getString(R.string.nativead_video_play)
      videoController.videoLifecycleCallbacks = getVideoCallbacks()
    } else {
      // If the main asset is a image, enable the UI.
      binding.textVideoStatus.text = getString(R.string.nativead_video_none)
    }
  }

  private fun getVideoCallbacks(): VideoController.VideoLifecycleCallbacks {
    return object : VideoController.VideoLifecycleCallbacks {
      override fun onVideoStart() {
        binding.textVideoStatus.text = getString(R.string.nativead_video_started)
        customControls?.onVideoStart()
        super.onVideoStart()
      }

      override fun onVideoEnd() {
        // Publishers should allow native ads to complete video playback before
        // refreshing or replacing them with another ad in the same UI location.
        binding.textVideoStatus.text = getString(R.string.nativead_video_ended)
        customControls?.onVideoEnd()
        super.onVideoEnd()
      }

      override fun onVideoPause() {
        binding.textVideoStatus.text = getString(R.string.nativead_video_pause)
        customControls?.onVideoPause()
        super.onVideoPause()
      }

      override fun onVideoPlay() {
        binding.textVideoStatus.text = getString(R.string.nativead_video_play)
        customControls?.onVideoPlay()
        super.onVideoPlay()
      }

      override fun onVideoMute(muted: Boolean) {
        customControls?.onVideoMute(muted)
        super.onVideoPlay()
      }
    }
  }

  private fun setUIEnabled(enableUI: Boolean) {
    isUIEnabled = enableUI
    updateUI()
  }

  private fun updateUI() {
    activity?.runOnUiThread {
      binding.refreshAdButton.isEnabled = isUIEnabled
      binding.checkRequestVideo.isEnabled = isUIEnabled
      binding.checkStartMuted.isEnabled = isUIEnabled && binding.checkRequestVideo.isChecked
    }
  }

  private companion object {
    // Sample native image ad unit ID.
    const val IMAGE_AD_UNIT_ID = "ca-app-pub-3940256099942544/2247696110"
    // Sample native video ad unit ID.
    const val VIDEO_AD_UNIT_ID = "ca-app-pub-3940256099942544/1044960115"
  }
}
